### PR TITLE
Fixed import of packages from requirement files with UTF-16 encoding

### PR DESCRIPTION
**Pull request by francescocaponio**
Solves #1386

### CODE REVIEW

**Review by ichard26 — COMMENTED** [review]
This doesn't seem to do anything? It's just opening the file twice without saving the read contents anywhere.

It's probably better to just loudly error in the logs if the filesystem/OS default encoding doesn't work. Maybe we could be smarter about which one we use, but that's beyond my expertise.

**Comment by francescocaponio** [comment]
Actually I searched a lot and there is no way to know before opening a file which is the encoding.

The try except block tries to open it first with UTF-8 encoding (the more common), if it fails with UnicodeDecode Exception, it opens it with UTF-16, maybe the UTF-16 open in the except block is unuseful, since it will be later opened again to be read. That could be avoided or used in another try except block, if it would be the case in the future (I would hope not). 

Windows powershell exports the requirements file in UTF-16, while bash does it in UTF-8.

As I was explaining ih the #1386, bandersnatch process is quite unattended, I don't want it to be stuck if a wrong encoded file goes in the requirements folder.

**Review by cooperlees — CHANGES_REQUESTED** [review]
Thanks for attempting a PR. I commented on the issue and I feel we should just use [charset-normalizer](https://pypi.org/project/charset-normalizer/) and detect the encoding and default to `utf-8` if it fails to detect so we support all encoding that `pip` does with requirements files.

**Comment by francescocaponio** [comment]
I didn't think about PEP rules and actually never noticed I had some UTF-16 encoded requirements until I provided them to bandersnatch. 
Pip installs them without any problem:
image
pip uses the function auto_decode to open files: https://github.com/pypa/pip/blob/61e85bf27996a9a73170e353374f13b742ad2058/src/pip/_internal/utils/encoding.py#L20
maybe we can use the same too?

**Comment by cooperlees** [comment]
This is fine that windows does this and pip supports it. Like add pip like support here to bandersnatch please and that will save us ever having to come back to it. I'm all for adding it, but i think it's worth the little bit of extra effort. Thank for diffing in and working it all out.

**Comment by francescocaponio** [comment]
Ok, sorry to bother you.
I have the edited version, but I'm unable to pass the pre-commit scripts. Can you help me?
The problem is happening here:
image
black adds a whitespace, and flake8 is unhappy about it:
image

Did I do something wrong?

BTW pytest is passing.

File: .flake8 [hunk 1]
```diff
@@ -4,7 +4,7 @@ max_line_length = 88
 # E722 is a duplicate of B001.
 # P207 is a duplicate of B003.
 # W503 is against PEP8
-ignore = E722, P207, W503
+ignore = E722, P207, W503, E203
```
Comment (cooperlees) [inline .flake8:7]: Thanks - Wonder how we never hit this ... Just must have never done it.

File: CHANGES.md [hunk 2]
```diff
@@ -7,6 +7,10 @@
 - Move Docker to build in 3.11 `PR #1341`
 - Add "--force-check" parameter to runner.py `PR #1347`
 
+## Bug Fixes
+
+- Fixed import of packages from requirement files with UTF-16 encoding `PR #1387`
```
Comment (cooperlees) [inline CHANGES.md:12]:
```suggestion
- Fixed reading requirement files with UTF-16 encoding supporting all encoding pip does `PR #1387`
```
We kind of do more than just UTF-16 now ... But I can fix later if you don't get to it.

**Review by cooperlees — APPROVED** [review]
LGTM - Thanks. This will indeed support all pip supports.

I might look into or open a follow up issues to see if we can remove this copy pasted code and move to a library that will be kept up to date for us. But this is fine for now and does more than only adding UTF-16

**Pull request merged**